posts/python_sql_engine.md

```diff
@@ -0,0 +1,187 @@
+
+Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that comments can be preserved.
+
+## Parsing
+
+Once a SQL statement is tokenized, we don't need to worry about white space and other formatting, so it's easier to work with. We can now convert the list of tokens into an AST. The SQLGlot [parser](https://github.com/tobymao/sqlglot/blob/main/sqlglot/parser.py) is a hand written [recursive descent](https://en.wikipedia.org/wiki/Recursive_descent_parser) parser.
+
+Similar to the tokenizer, it consumes the tokens sequentially but it uses a recursive algorithm instead. The tokens are converted into a single AST node which presents the SQL query. The SQLGlot parser was designed to support various dialects so it contains many options for overriding parsing functionality.
+
+![Parser Output](python_sql_engine_images/parser.png)
+
+The AST is a generic representation of a given SQL query. Each dialect can override or implement its own generator that can convert an AST object into syntatically correct SQL.
+
+## Optimizing
+
+Once we have our AST, we need to simplify it and put it into a consistent form to actually run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.
+
+1. I wanted a way to generate 'canonical sql'. Given two SQL statements, I wanted to be able to identify if they are semantically equivalent. For example `select 2` and `select 1 + 1` are functionally the same even though the SQL query is the same. Having the optimizer output SQL would allow me to convert 1 + 1 into 2.
+
+2. It's easier to debug and [validate](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/optimizer) the optimizations when the input and output are both SQL.
+
+I've yet to find another engine that takes this approach, but I'm quite happy with this decision. The optimizer currently does not perform any "physical optimizations" like join reordering. Those are left to the execution layer as additional statistics and information could become relevant.
+
+![Optimizer Output](python_sql_engine_images/optimizer.png)
+
+The optimizer currently has [17 rules](https://github.com/tobymao/sqlglot/tree/main/sqlglot/optimizer). Each of these rules is applied, transforming the AST in place. The combination of these rules creates "canonical" sql which can then be more easily converted into a logical plan and then executed.
+
+Some example rules are:
+
+### qualify\_tables and qualify_columns
+- Adds all db/catalog qualifiers to tables and forces an alias.
+- Ensure each column is unambiguous and expand stars.
+
+```sql
```

Comment:
We have a mix of inline code blocks and images.

Where do you plan to post this? Will it support inline code blocks with syntax highlighting?

Reply:
i plan to leave it in the mark down